File: backend/api/src/services/promptOverrides.ts
/**
 * Prompt override utilities.
 * 
 * Handles merging tenant-specific prompt overrides with the system defaults.
 * System defaults are typically imported from `@config/prompts`.
 */
import { ToolChoice, ToolConfig } from "../utils/types";
import { db } from "../utils/db";
import { env } from "../utils/env";

export const PROMPT_OVERRIDE_KEYS = [
  "workflow_generation",
  "workflow_step_generation",
  "workflow_edit",
  "workflow_instructions_refine",
  "template_html_generation",
  "template_metadata_generation",
  "html_patch",
  "form_field_generation",
  "form_css_generation",
  "form_css_refine",
  "execution_step_edit",
  "file_search_assistant",
  "file_search_simple",
  "styled_html_generation",
  "image_prompt_planner",
  "shell_tool_loop_default",
  "workflow_ideation",
  "workflow_ideation_followup",
] as const;

export type PromptOverrideKey = typeof PROMPT_OVERRIDE_KEYS[number];

export type PromptOverride = {
  enabled?: boolean;
  instructions?: string;
  prompt?: string;
  model?: string;
  tools?: ToolConfig[];
  tool_choice?: ToolChoice;
  service_tier?: string;
  reasoning_effort?: string;
  text_verbosity?: string;
};

export type PromptOverrides = Record<string, PromptOverride>;

const USER_SETTINGS_TABLE = env.userSettingsTable;

const coercePromptText = (value: unknown): string | undefined => {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  if (!trimmed) return undefined;
  return value;
};

export const normalizePromptOverrides = (
  raw: unknown,
): PromptOverrides | undefined => {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    return undefined;
  }

  const entries = Object.entries(raw as Record<string, unknown>);
  if (entries.length === 0) return {};

  const normalized: PromptOverrides = {};
  for (const [key, value] of entries) {
    if (!value || typeof value !== "object" || Array.isArray(value)) {
      continue;
    }
    const override = value as Record<string, unknown>;
    const instructions = coercePromptText(override.instructions);
    const prompt = coercePromptText(override.prompt);
    const enabled =
      typeof override.enabled === "boolean" ? override.enabled : undefined;
    const model = coercePromptText(override.model);
    const tools = Array.isArray(override.tools)
      ? (override.tools as ToolConfig[])
      : undefined;
    const tool_choice = coercePromptText(override.tool_choice) as
      | ToolChoice
      | undefined;
    const service_tier = coercePromptText(override.service_tier);
    const reasoning_effort = coercePromptText(override.reasoning_effort);
    const text_verbosity = coercePromptText(override.text_verbosity);

    if (
      instructions ||
      prompt ||
      enabled !== undefined ||
      model ||
      tools ||
      tool_choice ||
      service_tier ||
      reasoning_effort ||
      text_verbosity
    ) {
      normalized[key] = {
        ...(enabled !== undefined ? { enabled } : {}),
        ...(instructions ? { instructions } : {}),
        ...(prompt ? { prompt } : {}),
        ...(model ? { model } : {}),
        ...(tools ? { tools } : {}),
        ...(tool_choice ? { tool_choice } : {}),
        ...(service_tier ? { service_tier } : {}),
        ...(reasoning_effort ? { reasoning_effort } : {}),
        ...(text_verbosity ? { text_verbosity } : {}),
      };
    }
  }

  return normalized;
};

export const applyPromptTemplate = (
  template: string | undefined,
  variables: Record<string, string | undefined>,
): string | undefined => {
  if (!template) return template;

  return template.replace(/\{\{\s*([a-zA-Z0-9_.-]+)\s*\}\}/g, (match, key) => {
    const value = variables[key];
    if (value === undefined || value === null) {
      return match;
    }
    return String(value);
  });
};

export const resolvePromptOverride = ({
  key,
  defaults,
  overrides,
  variables = {},
}: {
  key: PromptOverrideKey;
  defaults: PromptOverride;
  overrides?: PromptOverrides;
  variables?: Record<string, string | undefined>;
}): PromptOverride => {
  const override = overrides?.[key];
  const isEnabled = override?.enabled !== false;

  const instructions =
    isEnabled && override?.instructions
      ? override.instructions
      : defaults.instructions;
  const prompt =
    isEnabled && override?.prompt ? override.prompt : defaults.prompt;
  
  const model = isEnabled && override?.model ? override.model : defaults.model;
  const tools = isEnabled && override?.tools ? override.tools : defaults.tools;
  const tool_choice =
    isEnabled && override?.tool_choice
      ? override.tool_choice
      : defaults.tool_choice;
  const service_tier =
    isEnabled && override?.service_tier
      ? override.service_tier
      : defaults.service_tier;
  const reasoning_effort =
    isEnabled && override?.reasoning_effort
      ? override.reasoning_effort
      : defaults.reasoning_effort;
  const text_verbosity =
    isEnabled && override?.text_verbosity
      ? override.text_verbosity
      : defaults.text_verbosity;

  return {
    instructions: applyPromptTemplate(instructions, variables),
    prompt: applyPromptTemplate(prompt, variables),
    model,
    tools,
    tool_choice,
    service_tier,
    reasoning_effort,
    text_verbosity,
  };
};

export const getPromptOverridesFromSettings = (
  settings?: Record<string, unknown> | null,
): PromptOverrides | undefined => {
  if (!settings) return undefined;
  return normalizePromptOverrides(settings.prompt_overrides);
};

export const getPromptOverridesForTenant = async (
  tenantId: string,
): Promise<PromptOverrides | undefined> => {
  if (!USER_SETTINGS_TABLE) return undefined;

  const settings = await db.get(USER_SETTINGS_TABLE, { tenant_id: tenantId });
  return getPromptOverridesFromSettings(settings);
};

File: backend/worker/services/prompt_overrides.py
"""Prompt override utilities for worker-side AI calls."""
from __future__ import annotations

import re
from typing import Any, Dict, Optional


def _coerce_prompt_text(value: Any) -> Optional[str]:
    if not isinstance(value, str):
        return None
    if not value.strip():
        return None
    return value


def normalize_prompt_overrides(raw: Any) -> Optional[Dict[str, Dict[str, Any]]]:
    if not isinstance(raw, dict):
        return None
    normalized: Dict[str, Dict[str, Any]] = {}
    for key, value in raw.items():
        if not isinstance(value, dict):
            continue
        instructions = _coerce_prompt_text(value.get("instructions"))
        prompt = _coerce_prompt_text(value.get("prompt"))
        enabled = value.get("enabled")
        enabled_flag = enabled if isinstance(enabled, bool) else None
        if instructions or prompt or enabled_flag is not None:
            normalized[key] = {}
            if enabled_flag is not None:
                normalized[key]["enabled"] = enabled_flag
            if instructions:
                normalized[key]["instructions"] = instructions
            if prompt:
                normalized[key]["prompt"] = prompt
    return normalized


def apply_prompt_template(
    template: Optional[str], variables: Dict[str, Optional[str]]
) -> Optional[str]:
    if not template:
        return template

    pattern = re.compile(r"\{\{\s*([a-zA-Z0-9_.-]+)\s*\}\}")

    def replace(match: re.Match[str]) -> str:
        key = match.group(1)
        value = variables.get(key)
        if value is None:
            return match.group(0)
        return str(value)

    return pattern.sub(replace, template)


def resolve_prompt_override(
    key: str,
    defaults: Dict[str, Any],
    overrides: Optional[Dict[str, Dict[str, Any]]],
    variables: Optional[Dict[str, Optional[str]]] = None,
) -> Dict[str, Any]:
    override = overrides.get(key) if overrides else None
    enabled = True
    if isinstance(override, dict) and isinstance(override.get("enabled"), bool):
        enabled = bool(override.get("enabled"))

    instructions = defaults.get("instructions")
    prompt = defaults.get("prompt")
    model = defaults.get("model")
    tools = defaults.get("tools")
    tool_choice = defaults.get("tool_choice")
    service_tier = defaults.get("service_tier")
    reasoning_effort = defaults.get("reasoning_effort")
    text_verbosity = defaults.get("text_verbosity")

    if enabled and isinstance(override, dict):
        instructions = override.get("instructions") or instructions
        prompt = override.get("prompt") or prompt
        model = override.get("model") or model
        tools = override.get("tools") or tools
        tool_choice = override.get("tool_choice") or tool_choice
        service_tier = override.get("service_tier") or service_tier
        reasoning_effort = override.get("reasoning_effort") or reasoning_effort
        text_verbosity = override.get("text_verbosity") or text_verbosity

    resolved_vars = variables or {}
    return {
        "instructions": apply_prompt_template(instructions, resolved_vars),
        "prompt": apply_prompt_template(prompt, resolved_vars),
        "model": model,
        "tools": tools,
        "tool_choice": tool_choice,
        "service_tier": service_tier,
        "reasoning_effort": reasoning_effort,
        "text_verbosity": text_verbosity,
    }


def get_prompt_overrides(
    db_service: Optional[Any],
    tenant_id: Optional[str],
) -> Optional[Dict[str, Dict[str, Any]]]:
    if not db_service or not tenant_id:
        return None
    try:
        settings = db_service.get_settings(tenant_id) or {}
    except Exception:
        return None
    return normalize_prompt_overrides(settings.get("prompt_overrides"))
